### PR TITLE
feat: global search categories suggestions 

### PR DESCRIPTION
**Pull request by sagarnaikjuspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
global search categories suggestions 
Screenshot 2024-11-13 at 11 38 53 PM

## How did you test it?
u should be able to pass multiple query in global search through filters 

## Where to test it?
Test cases
1. Valid Scenarios:
- Single predefined filter.
- Multiple predefined filters.
- Single free-text search.
- Combination of predefined filters and single free-text search.

2. Invalid Scenarios:
- Multiple free-text searches.
- Incorrect filter syntax.
- Undefined filters treated as free-text.

3. Edge Cases:
- Empty input.
- Extremely long input strings.
- Special characters in input.

- [x] INTEG
- [x] SANDBOX
- [x] PROD

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 1]
```diff
@@ -272,14 +267,27 @@ let getDefaultResult = searchText => {
   }
 }
 
+let getDefaultOption = searchText => {
+  open GlobalSearchTypes
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:271]: can you open `GlobalSearchTypes` at the top in the `GlobalSearchBarUtils` file

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 2]
```diff
@@ -272,14 +267,27 @@ let getDefaultResult = searchText => {
...
+  {
+    texts: ["Show all results for"->JSON.Encode.string, searchText->JSON.Encode.string],
+    redirect_link: `/search?query=${searchText}`->JSON.Encode.string,
+  }
+}
+
+let getAllOptions = (results: array<GlobalSearchTypes.resultType>) => {
+  open GlobalSearchTypes
+
+  []->Array.concatMany(results->Array.map(item => item.results))
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:281]: can you try using `Array.flatMap`

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 3]
```diff
@@ -300,3 +308,237 @@ let generateSearchBody = (~searchText, ~merchant_id) => {
...
+      source: "BATCH",
+    }
+    AnalyticsUtils.filterBody(filterBodyEntity)
+  }->Identity.genericTypeToJson
+
+let generateFilter = (queryArray: array<string>) => {
+  open LogicUtils
+  let filter = Dict.make()
+  queryArray->Array.forEach(query => {
+    let arr =
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:451]: can you please update variable name

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 4]
```diff
@@ -300,3 +308,237 @@ let generateSearchBody = (~searchText, ~merchant_id) => {
...
+      query
+      ->String.split(":")
+      ->Array.filter(query => {
+        let queryText = query->String.trim
+        queryText->isNonEmptyString
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:456]: Array.filter(query => query->String.trim->isNonEmptyString)

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 5]
```diff
@@ -300,3 +308,237 @@ let generateSearchBody = (~searchText, ~merchant_id) => {
...
+    let query = query->Array.concat([("query", queryText.contents->JSON.Encode.string)])
+
+    query->Dict.fromArray
+  }
+
+  body
+}
+
+let validateQuery = searchQuery => {
+  open LogicUtils
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:514]: can you open `LogicUtils` once at the top of file

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res [hunk 6]
```diff
@@ -300,3 +308,237 @@ let generateSearchBody = (~searchText, ~merchant_id) => {
...
+
+  let freeTextCount = ref(0)
+
+  searchQuery
+  ->String.split(" ")
+  ->Array.filter(query => {
+    let queryText = query->String.trim
+    queryText->isNonEmptyString
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarUtils.res:522]: can remove extra variable

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarHelper.res [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+  }
+}
+
+module OptionWrapper = {
+  @react.component
+  let make = (~index, ~value, ~children, ~selectedOption, ~redirectOnSelect) => {
+    let activeClass = value == selectedOption ? "bg-gray-100 rounded-lg p-2 group items-center" : ""
+    <div
+      onClick={_ => {
+        value->redirectOnSelect
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarHelper.res:71]: can remove curly braces

File: src/screens/Analytics/GlobalSearch/GlobalSearchBarHelper.res [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+        | Others => React.null
+        }
+      }
+    </RenderIf>
+  }
+}
+
+module SearchResultsComponent = {
+  open GlobalSearchTypes
+  open LogicUtils
```
Comment (gitanjli525) [inline src/screens/Analytics/GlobalSearch/GlobalSearchBarHelper.res:163]: can we open `GlobalSearchTypes` and `LogicUtils` at the top of the file

**Review by gitanjli525 — DISMISSED** [review]

**Review by JeevaRamu0104 — APPROVED** [review]

**Review by JeevaRamu0104 — APPROVED** [review]

**Pull request merged**